Marlin/src/module/planner.cpp

```diff
@@ -794,22 +794,27 @@ block_t* Planner::get_current_block() {
  */
 void Planner::calculate_trapezoid_for_block(block_t * const block, const_float_t entry_factor, const_float_t exit_factor) {
 
-  uint32_t initial_rate = CEIL(block->nominal_rate * entry_factor),
-           final_rate = CEIL(block->nominal_rate * exit_factor); // (steps per second)
+  uint32_t initial_rate = LROUND(block->nominal_rate * entry_factor),
+           final_rate = LROUND(block->nominal_rate * exit_factor); // (steps per second)
 
-  // Limit minimal step rate (Otherwise the timer will overflow.)
+  // Legacy check against supposed timer overflow. However Stepper::calc_timer_interval() already
+  // should protect against it. But removing results in less smooth motion for switching direction
+  // moves. This is because the current discrete stepping math diverges from physical motion under
+  // constant acceleration when acceleration_steps_per_s2 is large compared to initial/final_rate.
   NOLESS(initial_rate, uint32_t(MINIMAL_STEP_RATE));
   NOLESS(final_rate, uint32_t(MINIMAL_STEP_RATE));
+  NOMORE(initial_rate, block->nominal_rate);
+  NOMORE(final_rate, block->nominal_rate);
```

Comment:
Are these fixing observed issues, or were these just added to protect against the possibility that entry/exit factors > 1.0 could be passed-in?

Reply:
To address the case where `MINIMAL_STEP_RATE` is greater than `nominal_rate`. In this case the existing:
```
NOLESS(initial_rate, uint32_t(MINIMAL_STEP_RATE));
NOLESS(final_rate, uint32_t(MINIMAL_STEP_RATE));
```
make `initial/final_rate` greater than `nominal_rate` and the math becomes all wonky.